File: server/models/site.py
from typing import Optional
from pydantic import BaseModel, ConfigDict
from datetime import datetime

class DailyCrowdnessPrediction(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,        
    )
    
    site_id: int
    h0: int = 0
    h1: int = 0
    h2: int = 0
    h3: int = 0
    h4: int = 0
    h5: int = 0
    h6: int = 0
    h7: int = 0
    h8: int = 0
    h9: int = 0
    h10: int = 0
    h11: int = 0
    h12: int = 0
    h13: int = 0
    h14: int = 0
    h15: int = 0
    h16: int = 0
    h17: int = 0
    h18: int = 0
    h19: int = 0
    h20: int = 0
    h21: int = 0
    h22: int = 0
    h23: int = 0

class WeeklyCrowdnessPrediction(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True
    )
    
    site_id: int
    Monday: int = 0
    Tuesday: int = 0
    Wednesday: int = 0
    Thursday: int = 0
    Friday: int = 0
    Saturday: int = 0
    Sunday: int = 0

class Site(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
    )

    site_id: Optional[int] = None
    site_name: str
    site_name_short: str
    site_desc: Optional[str] = None
    site_url: Optional[str] = None
    site_banner_url: Optional[str] = None

    daily_prediction: Optional[DailyCrowdnessPrediction] = None
    weekly_prediction: Optional[WeeklyCrowdnessPrediction] = None

class RawCrowdnessReading(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True
    )
    
    id: Optional[int] = None
    time: datetime
    site_id: int
    crowdness: int

File: client/app/models/site.ts
// Base model for Site and its related predictions

/**
 * Interface representing the structure of daily crowdness predictions for a site.
 * Corresponds to the SQLAlchemy 'DailyCrowdnessPrediction' model.
 */
export interface DailyCrowdnessPrediction {
  site_id: number; // Foreign key to sites.site_id, also primary key
  h0: number;
  h1: number;
  h2: number;
  h3: number;
  h4: number;
  h5: number;
  h6: number;
  h7: number;
  h8: number;
  h9: number;
  h10: number;
  h11: number;
  h12: number;
  h13: number;
  h14: number;
  h15: number;
  h16: number;
  h17: number;
  h18: number;
  h19: number;
  h20: number;
  h21: number;
  h22: number;
  h23: number;
  // The 'sites' relationship back to Site is represented by this record being nested
  // or linked within a Site object in TypeScript, typically via 'daily_prediction'.
}

/**
 * Interface representing the structure of weekly crowdness predictions for a site.
 * Corresponds to the SQLAlchemy 'WeeklyCrowdnessPrediction' model.
 */
export interface WeeklyCrowdnessPrediction {
  site_id: number; // Foreign key to sites.site_id, also primary key
  Monday: number;
  Tuesday: number;
  Wednesday: number;
  Thursday: number;
  Friday: number;
  Saturday: number;
  Sunday: number;
  // The 'sites' relationship back to Site is represented by this record being nested
  // or linked within a Site object in TypeScript, typically via 'weekly_prediction'.
}
export interface WaveQuality {
  site_id: number;
  wave_height: number;
  wave_speed: number;
  wave_direction: string;
  temperature: number;
}

/**
 * Interface representing a tourism site or location.
 * Corresponds to the SQLAlchemy 'Site' model.
 */
export interface Site {
  site_id: number;
  site_name: string;
  site_name_short: string;
  site_desc: string | null;
  site_url: string | null;
  site_banner_url: string | null;
  daily_prediction?: DailyCrowdnessPrediction | null; // Relationship: one-to-one or null
  weekly_prediction?: WeeklyCrowdnessPrediction | null; // Relationship: one-to-one or null
  // raw_crowdness_readings could be an array here if you were to define that relationship
  // on the Site model in SQLAlchemy (e.g., readings: Mapped[list["RawCrowdnessReading"]])
  wave_quality?: WaveQuality | null;
}
